Source/WTF/wtf/NativePromise.h

```diff
@@ -1466,23 +1466,37 @@ class NativePromiseRequest final {
     RefPtr<typename PromiseType::Request> m_request;
 };
 
+template<typename S, typename E>
+Ref<NativePromise<S, E>> createSettledPromise(Expected<S, E>&& result)
+{
+    return NativePromise<S, E>::createAndSettle(WTFMove(result));
+}
+
 // Invoke a function object (e.g., lambda) asynchronously.
 // Returns a promise that the function should eventually resolve or reject once the original promise returned by the lambda
 // is itself resolved or rejected.
 template<typename Function>
 static auto invokeAsync(SerialFunctionDispatcher& targetQueue, Function&& function, const Logger::LogSiteIdentifier& callerName = DEFAULT_LOGSITEIDENTIFIER)
 {
     static_assert(!std::is_lvalue_reference_v<Function>, "Function object must not be passed by lvalue-ref (to avoid unplanned copies); WTFMove() the object.");
-    static_assert(IsConvertibleToNativePromise<typename RemoveSmartPointer<decltype(function())>::type>, "Function object must return Ref<NativePromise>");
     using ReturnType = typename RemoveSmartPointer<decltype(function())>::type;
+    static_assert(IsConvertibleToNativePromise<ReturnType> || IsExpected<ReturnType>::value, "Function object must return Ref<NativePromise> or Expected<T, F>");
```

Comment:
You always `RemoveSmartPointer` before you test, this could misleadingly accept `NativePromise` (without `Ref`), `RefPtr<NativePromise>`, or `Ref<Expected>` (and explode with a less-friendly error below).
So you may have to be more specific, something like:
```
using ReturnType = decltype(function());
constexpr bool returnTypeIsRef = IsSmartRef<ReturnType>::value;
using ReturnTypeNoRef = typename RemoveSmartPointer<ReturnType>::type;
static_assert(returnTypeIsRef && IsConvertibleToNativePromise< ReturnTypeNoRef> || IsExpected<ReturnType>::value, "Function object must return Ref<NativePromise> or Expected");
... change the other tests&types accordingly
```

Reply:
That it could return NativePromise instead of Ref<NativePromise> is an issue indeed, however it would immediately be wrapped in a Ref<NativePromise.

Ref<Expected> wouldn't compile as it's not ref-counted anyway. but I see your point.